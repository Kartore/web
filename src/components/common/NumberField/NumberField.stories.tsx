import { NumberField } from '.';
import { Meta, StoryObj } from '@storybook/react';

const meta = {
  component: NumberField,
} satisfies Meta<typeof NumberField>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {},
};
