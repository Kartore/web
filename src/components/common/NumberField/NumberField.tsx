import type { FC } from 'react';
import { useRef } from 'react';

import type { AriaNumberFieldProps } from 'react-aria';
import { useLocale, useNumberField } from 'react-aria';
import { useNumberFieldState } from 'react-stately';

import { Button } from '~/components/common/Button';
import { cn } from '~/utils/tailwindUtil';

export type NumberFieldProps = {
	className?: string;
	showButton?: boolean;
} & AriaNumberFieldProps;

export const NumberField: FC<NumberFieldProps> = ({
	className,
	showButton,
	label,
	...props
}) => {
	const { locale } = useLocale();
	const state = useNumberFieldState({ ...props, label, locale });
	const ref = useRef(null);
	const {
		labelProps,
		inputProps,
		groupProps,
		decrementButtonProps,
		incrementButtonProps,
		descriptionProps,
		errorMessageProps,
		isInvalid,
		validationErrors,
	} = useNumberField({ ...props, label }, state, ref);

	return (
		<div
			className={cn('flex flex-row items-center justify-between', className)}
		>
			<label
				{...labelProps}
				className={cn(
					'font-semibold text-gray-600 text-sm',
					labelProps.className,
				)}
			>
				{label}
			</label>
			<div {...groupProps} className={'w-1/2'}>
				<input
					{...inputProps}
					ref={ref}
					className={cn(
						'w-full rounded border-none bg-gray-100 px-2 py-1 font-semibold text-sm transition-colors hover:bg-gray-200 focus-visible:bg-gray-200 focus-visible:outline-0',
						inputProps.className,
					)}
				/>
				{showButton ? (
					<>
						<Button {...decrementButtonProps}>-</Button>
						<Button {...incrementButtonProps}>+</Button>
					</>
				) : null}
			</div>
			{props.description && (
				<div
					{...descriptionProps}
					className={cn('text-xs', descriptionProps.className)}
				>
					{props.description}
				</div>
			)}
			{isInvalid && (
				<div
					{...errorMessageProps}
					className={cn('text-red text-xs', errorMessageProps.className)}
				>
					{validationErrors.join(' ')}
				</div>
			)}
		</div>
	);
};

NumberField.displayName = 'NumberField';
