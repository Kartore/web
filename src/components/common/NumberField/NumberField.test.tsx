import { describe } from 'vitest';
import { render } from '@testing-library/react';
import { composeStories } from '@storybook/react';

import * as Stories from './NumberField.stories';

const { Default } = composeStories(Stories);

describe('Component: NumberField', () => {
  describe('Snapshot', () => {
    it('Default', () => {
      const { asFragment } = render(<Default />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
