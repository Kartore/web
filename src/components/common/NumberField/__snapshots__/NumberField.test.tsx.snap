// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: NumberField > Snapshot > Default 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row items-center justify-between"
  >
    <label
      class="text-sm font-semibold text-gray-600"
    />
    <div
      class="w-1/2"
      role="group"
    >
      <input
        aria-roledescription="Number field"
        autocomplete="off"
        autocorrect="off"
        class="w-full rounded border-none bg-gray-100 py-1 px-2 text-sm font-semibold transition-colors hover:bg-gray-200 focus-visible:bg-gray-200 focus-visible:outline-0"
        id="react-aria-:r0:"
        inputmode="numeric"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
